src/charm_state.py

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""State of the Charm."""
+import itertools
+import typing
+
+# pydantic is causing this no-name-in-module problem
+from pydantic import (  # pylint: disable=no-name-in-module,import-error
+    BaseModel,
+    Extra,
+    Field,
+    ValidationError,
+    validator,
+)
+
+from exceptions import CharmConfigInvalidError
+
+if typing.TYPE_CHECKING:
+    from charm import SynapseCharm
+
+
+KNOWN_CHARM_CONFIG = (
+    "server_name",
+    "report_stats",
+)
+
+
+class SynapseConfig(BaseModel):  # pylint: disable=too-few-public-methods
+    """Represent Synapse builtin configuration values.
+
+    Attrs:
+        server_name: server_name config.
+        report_stats: report_stats config.
+    """
+
+    server_name: str | None = Field(None, min_length=0)
+    report_stats: str | None = Field(None)
+
+    class Config:  # pylint: disable=too-few-public-methods
+        """Config class.
+
+        Attrs:
+            extra: extra configuration.
+        """
+
+        extra = Extra.allow
+
+    @validator("server_name")
+    # no-self-argument disabled due to cls usage as expected by validator
+    def check_server_name_empty(cls, value: str) -> str:  # pylint: disable=no-self-argument
+        """Check if server name is empty.
+
+        Args:
+            value: server_name value.
+
+        Raises:
+            ValueError: Raised if server_name is empty.
+
+        Returns:
+            server_name value.
+        """
+        if not value:
+            raise ValueError("The server_name is empty, please review your charm configuration")
+        return value
+
+    @validator("report_stats")
+    @classmethod
+    def to_yes_or_no(cls, value: str) -> str:
```

Comment:
I think if the `report_stats` is defined as `report_stats: typing.Literal["yes", "no"]`, then the validation function here is not needed.

Reply:
Gotcha, thanks.